### ISSUE TITLE
Check failed: index[i] < (int)tableSize

### ISSUE THREAD

**Issue by jamestang0219** [baidu/Paddle#333]
Hello, I have trained a LSTM model, and now I'm using PaddlePaddle's prediction python api to test the samples, but durning the prediction, the error occurred. Check failed: index[i] < (int)tableSize, does it means the value is out of boundary? The same sample can be used to train this model, but cannot be predicted, why? thanks.
here is the log:
```
I1103 07:12:24.671309 100820 Util.cpp:144] commandline:  --use_gpu=0
I1103 07:12:24.671407 100820 Util.cpp:113] Calling runInitFunctions
I1103 07:12:24.672067 100820 Util.cpp:126] Call runInitFunctions done.
I1103 07:12:29.378867 100820 GradientMachine.cpp:123] Loading parameters from ../data/pass-00004-208/
F1103 07:13:14.820614 100820 Matrix.cpp:2317] Check failed: index[i] < (int)tableSize (332123 vs. 332123)
*** Check failure stack trace: ***
    @     0x7fa7924dcdaa  (unknown)
    @     0x7fa7924dcce4  (unknown)
    @     0x7fa7924dc6e6  (unknown)
    @     0x7fa7924df687  (unknown)
    @     0x7fa7934aca49  paddle::CpuMatrix::selectRowsImp<>()
    @     0x7fa7934aac66  paddle::CpuMatrix::selectRows()
    @     0x7fa7933e2b37  paddle::TableProjection::forward()
    @     0x7fa7933bd266  paddle::MixedLayer::forward()
    @     0x7fa793358a8c  paddle::NeuralNetwork::forward()
    @     0x7fa793307d25  _wrap_GradientMachine_forward
```

**Comment by jamestang0219** [comment]
whatever I used prediction api or "paddle train --job=test", got the same result.

**Comment by luotao1** [comment]
Do you use the same sample for train/test/predict?

**Comment by jamestang0219** [comment]
@luotao1 use the same sample to test(--job=test) and predict(python api)

**Comment by luotao1** [comment]
can you paste your model and dataprovider

**Comment by jamestang0219** [comment]
@luotao1 dataprovider:
`from paddle.trainer.PyDataProvider2 import *
import json

UNK_IDX = 0
dict_file = "./data/pure_dict"
word_dict = dict()
counter = 1
with open(dict_file, 'r') as f:
    for line in f:
        data = line.strip().split("\t")
        if len(data) !=2 :
            continue
        if int(data[1]) >= 50:
            word_dict[data[0].decode('utf-8').encode('utf-8')] = counter
            counter += 1

def initializer(settings, **kwargs):
    settings.input_types = [
        integer_value_sequence(len(word_dict)),
        integer_value(2)]

@provider(init_hook=initializer, cache=CacheType.CACHE_PASS_IN_MEM)
def process(settings, file_name):
    with open(file_name, 'r') as f:
        for line in f:
            data = json.loads(line.strip())
            label = 0
            if data['label']==-1:
                label=1
            words = data['words']
            word_slot = [word_dict.get(w, UNK_IDX) for w in words]
            yield word_slot, label


def predict_initializer(settings, **kwargs):
    settings.input_types = [
        integer_value(len(word_dict), seq_type=SequenceType.SEQUENCE)
    ]

@provider(init_hook=predict_initializer, should_shuffle=False)
def process_predict(settings, file_name):
    with open(file_name, 'r') as f:
        for line in f:
            data = json.loads(line.strip())
            words = data['words']
            word_slot = [word_dict.get(w, UNK_IDX) for w in words]
            yield word_slot`

**Comment by jamestang0219** [comment]
@luotao1 model conf
`from paddle.trainer_config_helpers import *

dict_file = "./data/google_pure_dict"
word_dict = dict()
counter = 1
with open(dict_file, 'r') as f:
    for line in f:
        data = line.strip().split("\t")
        if int(data[1]) >= 50:
            word_dict[data[0].decode('utf-8').encode('utf-8')] = counter
            counter += 1

is_predict = get_config_arg('is_predict', bool, True)
train_set = 'data/train.list' if not is_predict else None
test_set = 'data/test.list' if not is_predict else 'data/pred.list'
process = 'process' if not is_predict else 'process_predict'
define_py_data_sources2(train_list=train_set,
                        test_list=test_set,
                        module="dataprovider",
                        obj=process)

batch_size = 128 if not is_predict else 1
settings(
    batch_size=batch_size,
    learning_rate=2e-3,
    learning_method=AdamOptimizer(),
    regularization=L2Regularization(8e-4),
    gradient_clipping_threshold=25
)

bias_attr = ParamAttr(initial_std=0.,l2_rate=0.)

data = data_layer(name="word", size=len(word_dict))
emb = embedding_layer(input=data, size=128)
fc = fc_layer(input=emb, size=512,
              act=LinearActivation(),
              bias_attr=bias_attr,
              layer_attr=ExtraAttr(drop_rate=0.1))
lstm = lstmemory(input=fc, act=TanhActivation(),
                 bias_attr=bias_attr,
                 layer_attr=ExtraAttr(drop_rate=0.25))
lstm_pool = pooling_layer(input=lstm, pooling_type=MaxPooling())
output = fc_layer(input=lstm_pool, size=2,
                  bias_attr=bias_attr,
                  act=SoftmaxActivation())

if is_predict:
    maxid = maxid_layer(output)
    outputs([maxid, output])
else:
    label = data_layer(name="label", size=2)
    cls = classification_cost(input=output, label=label)
    auc = auc_evaluator(input=output, label=label)
    outputs(cls)`

**Comment by jamestang0219** [comment]
@luotao1 I found it's very werid. I've set test_period=1000, after training 1000 batches, it can SUCCESSFULLY test the validation samples. But after training all batches, I use the dumped model to test the same validation samples, the error occurred.

**Comment by luotao1** [comment]
是说启动一个train任务，在train的过程中，是可以正常test的；但是只启动test任务或predict任务，就core了？

**Comment by jamestang0219** [comment]
@luotao1 Yes, the same samples can be tested durning TRAIN JOB, but they cannot be tested or predicted in TEST JOB or PREDICT API

**Comment by luotao1** [comment]
你能用测试集训练一遍么，看看有没有问题

**Comment by jamestang0219** [comment]
@luotao1 process_predict was used in this place to define the data sources:
process = 'process' if not is_predict else 'process_predict', define_py_data_sources2(train_list=train_set,
                        test_list=test_set,
                        module="dataprovider",
                        obj=process)

and I've changed the pred.list to PREDICT all train samples, here is the test log:

`[INFO 2016-11-03 10:13:46,784 networks.py:960] The input order is [word]
[INFO 2016-11-03 10:13:46,785 networks.py:963] The output order is []
I1103 10:13:46.841385 107001 Trainer.cpp:148] trainer: in testing mode
I1103 10:13:46.841454 107001 Trainer.cpp:155] trainer mode: Testing
I1103 10:13:47.506799 107001 PyDataProvider2.cpp:219] loading dataprovider dataprovider::process_predict
I1103 10:13:50.471432 107001 GradientMachine.cpp:123] Loading parameters from model_desc/pass-00004
I1103 10:16:28.491477 107001 Tester.cpp:136]  Batch=100 samples=100 AvgCost=0
I1103 10:17:06.230001 107001 Tester.cpp:136]  Batch=200 samples=200 AvgCost=0
I1103 10:17:44.747331 107001 Tester.cpp:136]  Batch=300 samples=300 AvgCost=0
I1103 10:18:14.180492 107001 Tester.cpp:136]  Batch=400 samples=400 AvgCost=0
I1103 10:18:17.137424 107001 Tester.cpp:136]  Batch=500 samples=500 AvgCost=0
I1103 10:18:19.970542 107001 Tester.cpp:136]  Batch=600 samples=600 AvgCost=0
I1103 10:18:22.860966 107001 Tester.cpp:136]  Batch=700 samples=700 AvgCost=0
I1103 10:18:25.563513 107001 Tester.cpp:136]  Batch=800 samples=800 AvgCost=0
I1103 10:18:28.244895 107001 Tester.cpp:136]  Batch=900 samples=900 AvgCost=0
F1103 10:18:29.546298 107001 Matrix.cpp:2317] Check failed: index[i] < (int)tableSize (332123 vs. 332123)
*** Check failure stack trace: ***
    @     0x7fb070e57daa  (unknown)
    @     0x7fb070e57ce4  (unknown)
    @     0x7fb070e576e6  (unknown)
    @     0x7fb070e5a687  (unknown)
    @           0x6d96c9  paddle::CpuMatrix::selectRowsImp<>()
    @           0x6d5186  paddle::CpuMatrix::selectRows()
    @           0x601907  paddle::TableProjection::forward()
    @           0x5b54c6  paddle::MixedLayer::forward()
    @           0x67251c  paddle::NeuralNetwork::forward()
    @           0x6b576f  paddle::Tester::testOneBatch()
    @           0x6b6683  paddle::Tester::testOneBatchById()
    @           0x6b699f  paddle::Tester::testOnePass()
    @           0x6b6e6b  paddle::Tester::test()
    @           0x53b237  main
    @     0x7fb070063f45  (unknown)
    @           0x546585  (unknown)
    @              (nil)  (unknown)
/usr/local/bin/paddle: line 46: 107001 Aborted                 (core dumped) ${DEBUGGER} $MYDIR/../opt/paddle/bin/paddle_trainer ${@:2}`

**Comment by reyoung** [comment]
@jamestang0219 Should the data layer be size of `len(word_dict) + 1` because there is an UNK in data_layer and not in `word_dict `?

**Comment by jamestang0219** [comment]
@reyoung yes, just now I found this bug:P
`I1103 10:55:33.235699 111346 PyDataProvider2.cpp:219] loading dataprovider dataprovider::process_predict
332124
I1103 10:55:36.208869 111346 GradientMachine.cpp:123] Loading parameters from model_desc/pass-00004
[139809, 212822, 223124, 125812, 26902, 289091, 239996, 2422, 175863, 141806, 58449, 152461, 141463, 49586, 43203, 88369, 293854, 175863, 201399, 58449, 332123, 11587, 327063, 319765, 26902, 134230, 279057, 249392, 26918, 303608, 239996, 312313, 316237, 128779, 217547, 58449, 73209, 293379, 30158, 167530, 275732, 256515, 21078, 21631, 84923, 604, 281649, 101353, 88369, 150078, 72782, 202914, 175863, 271856, 139809, 239996, 263643, 58449, 35138, 196541, 95133, 176075, 1548, 195001, 72782, 32271, 58449, 13077, 211023, 160191, 225179, 220556, 175863, 328952, 293854, 165176, 26902, 58449, 161409, 88369, 201687, 293854, 130867, 169329, 11587, 182946, 167533, 33219, 239996, 58449, 39736, 374, 67478, 65011, 204775, 284011, 58449, 124454, 73022, 0]
F1103 10:55:36.438951 111346 Matrix.cpp:2317] Check failed: index[i] < (int)tableSize (332123 vs. 332123)
*** Check failure stack trace: ***
    @     0x7f3dfa596daa  (unknown)
    @     0x7f3dfa596ce4  (unknown)
    @     0x7f3dfa5966e6  (unknown)
    @     0x7f3dfa599687  (unknown)
    @           0x6d96c9  paddle::CpuMatrix::selectRowsImp<>()
    @           0x6d5186  paddle::CpuMatrix::selectRows()
    @           0x601907  paddle::TableProjection::forward()
    @           0x5b54c6  paddle::MixedLayer::forward()
    @           0x67251c  paddle::NeuralNetwork::forward()
    @           0x6b576f  paddle::Tester::testOneBatch()
    @           0x6b6683  paddle::Tester::testOneBatchById()
    @           0x6b699f  paddle::Tester::testOnePass()
    @           0x6b6e6b  paddle::Tester::test()
    @           0x53b237  main
    @     0x7f3df97a2f45  (unknown)
    @           0x546585  (unknown)
    @              (nil)  (unknown)
/usr/local/bin/paddle: line 46: 111346 Aborted                 (core dumped) ${DEBUGGER} $MYDIR/../opt/paddle/bin/paddle_trainer ${@:2}`

Thank you!

**Comment by reyoung** [comment]
you are welcome.
